engine/suspender.go

```diff
@@ -15,90 +15,104 @@ import (
 // it will read and write namespaces' annotations, and scale resources.
 func (eng *Engine) Suspender(ctx context.Context, cs *kubernetes.Clientset) {
 	eng.Mutex.Lock()
-	eng.RunningNamespacesList = make(map[string]time.Time)
 	eng.Logger.Info().Str("routine", "suspender").Msg("suspender started")
 	eng.Mutex.Unlock()
 
+	var id int
 	for {
+		// QUESTION: Why unlocking here ?
 		eng.Mutex.Unlock()
+
 		// wait for the next namespace to check
 		n := <-eng.Wl
 		start := time.Now()
 
-		// we create a sublogger to avoid "namespace" field duplication at each
-		// loop
+		// we create a sublogger to avoid "namespace" field duplication at each loop
 		eng.Mutex.Lock()
 		sLogger := eng.Logger.With().Str("routine", "suspender").Str("namespace", n.Name).Logger()
-		sLogger.Trace().Msgf("namespace %s received from watcher", n.Name)
-
-		// get the namespace state
-		dState := n.Annotations[eng.Options.Prefix+DesiredState]
+		sLogger.Debug().Msgf("namespace received from watcher")
 
 		/*
+			Step 1
+
 			This first switch-case statement will ensure that the namespace has a state set.
 			- if dState is empty, it means that it is the first time we see this namespace, so we
 			add the annotation with the state Running
 			- if dState is equal to Running or Suspended, the switch-case will do nothing.
 			- if dState ends in the default case, it means that the state has not been recognised, so
 			we have to error
 		*/
-
+		id = 1
```

Comment:
:thinking: adding an id like that increment "randomly" in the code does not seem to be along idea on the long run, I feel that it will be hard to maintain in the future
- when do we need to increment it?
- what is a "step"?

IMO it is better to add context to logging instead of just an int that does not provide much information by itself (in 6 months we will have to dig in the code to remember what `step 2` means).

Reply:
It is not "randomly" incremented, it is regarding the "steps", today only mentioned in the comments sections.

It adds context to me since we do kind of twice the work when switching on `dState`.
It helps me know where we are in the process while reading the logs and debugging.

I do not argue about the long-term maintenance effort and I agree that today, it is badly done.
But it is a starting point and I will be more than happy to make it better later by refactoring.

We definitively need to define clearer "steps" by making functions and keeping the core of `suspender` smaller.

Reply:
let's use string instead of an int, and use a (user|developer|debugger)-friendly, which describes the name of the step instead of the id.

`.Str("step", "check if ns has a state")`
`.Str("step", "update replicas numbers")`

(those are examples)